PublicMapping/districtbuilder-classic
# Replace Ansible provisioning with a `docker-compose` based deployment

## Overview

This PR replaces complicated Ansible tooling with `docker-compose` based deployments to simplify the deployment process and decouple application deployment from Azavea's AWS infrastructure. Now, users will be able to deploy the application on the infrastructure provider of their choice without being tied to our internal deployment scripts. Additionally, I added a `Deployment` section to the project `README` to separate the general deployment instructions from internal ones.

- Remove ansible provisioning files
    - The App server instance now installs dependencies and does pre-deploy setup using `cloud-init`.
- Add `docker-compose.production.yml`
    - `docker-compose.yml` defines the base architecture for the application
    - `docker-compose.overrides.yml` contains development environment-specific configurations (volume mounts, `postgres` container, etc).
    - `docker-compose.production.yml` contains Prod environment-specific configurations  (send all logs to syslog, `on-failure` restart policy)
- Remove unnecessary STRTA (`scripts/cibuid`, `scripts/cipublish`)
- Add remote `docker-compose` based deploy
    - Install `docker-compose` in Terraform container
    - Add additional AppServer security groups to allow Docker access from CI
    - Secure Docker by enabling TLS
    - Add a `--production` flag to development STRTA for cross-environment functionality
    - Modify `scripts/infra` to check if docker is running, then build images and start services directly on the AppServer instance
- Update deployment documentation

### Checklist

- [ x ] PR has a name that won't get you publicly shamed for vagueness

### Demo

https://va.districtbuilder.azavea.com

### Notes
- For more info on Docker over HTTPS, see [Docker Docs](https://docs.docker.com/engine/security/https/)
- Docker server TLS certificates are stored at `s3://districtbuilder-staging-config-us-east-1/docker_certs/server/{pa,va}.zip`. Certificates for other instances will need to be created manually and stored in the same location.
- Docker client TLS certificates are stored at `s3://districtbuilder-staging-config-us-east-1/docker_certs/client/client.zip`. These are general purpose and do not need to be created on a per-state basis.


## Testing Instructions

 * I modified `.travis.yml` to deploy this branch to staging. See [Travis Console](https://travis-ci.org/PublicMapping/DistrictBuilder/builds/373527103?utm_source=github_status&utm_medium=notification)
* See https://va.districtbuilder.azavea.com

Closes #542 
Closes #544
